### PR TITLE
Visualizations: Dynamically set any config (min, max, unit, color, thresholds) from query results

### LINKED ISSUE #159: Dynamic Grid Threshholds
Could Grid thresholds be fed dynamically as well? We feed some of our thresholds into graphite at work, it'd be useful to set the thresholds based on those metrics.

Cheers!

### LINKED ISSUE #3935: Feature Request: variables in X/Y-Max/Min
Let's say you have a set of network graphs for several servers. They all have different scale as Y-Max is set to auto. You can set it to 1073741824 ( 1 gigabit ) for all the graphs but it won't make much sense. Defining a max value with a template variable could be really helpful in this case.

### LINKED ISSUE #4929: [Feature Request] Gauge min/max value as variable
Hi,
It's will be nice if we could use a variable as min and max value in gauge panel.
In my case, I monitor utilization of data and have data formatted like this : 

```
Login | Space used | Quota
Foo   | 35 Gb      |50 Gb
Bar   | 65 Gb      |100 Gb 
...
```

So "space used" is the query result and "Quota" should be Max value of gauge.
I think that is there other usage of variable as min/max value.

Thank you very much for your great job !

### LINKED ISSUE #29497: Support of variable thresholds
**What would you like to be added**:
Currently, thresholds (e.g. in bar gauge charts) need to be predefined values. In v7.3.3 you could also enter characters. I tried to define query variables and use them for dynamic thresholds. Though this was working in principle, it was buggy & thresholds were not saved upon dashboard reload. It would be fantastic if variables could be set as thresholds.

**Why is this needed**:
Variable thresholds are more flexible than static ones.

### PR DESCRIPTION
**Pull request by torkelo**
Work in progress

* [x] Initial Rows to fields transform
* [ ] Improve thresholds handling 
* [x] Add support for more options (decimals, displayName, value mappings)
* [x] Rows to fields UI to select name and value field 
* [x] Value mapping support via new reducer "All values" that returns the array of values 
* [x] ConfigFromQuery transform (will reuse some parts of rows to fields) 
* [x] ConfigFromQuery editor UI 
* [x] Add option to change matcher & matcher options 
* [x] Add option to change reducer 
* [x] Make it possible to clear / ignore config fields 
* [x] Share field to config mapping UI  
* [x] Unit tests for RowsToFields transform & UI Editor 
* [x] Unit tests for ConfigFromQiery transform & UI Editor 
* [x] Unit tests for FieldToConfigMappingEditor 
* [x] Test dashboard for Rows To Fields transform
* [x] Test dashboard for Config from query transform  
* [x] Support for turning fields to labels in Rows to fields transform 
* [ ] Docs 

Follow-up
* Show what fields are overridden in the side options pane (Will be done in separate PR)
* Add beta / preview state feature to transforms (separate PR) 
* Improve transforms UI (show help/ escription after adding, disable single transform, disable all) 

Closes #159
Closes #29497 
Closes #4929 
Closes #3935 

Rows to fields 
Screenshot from 2021-07-08 11-23-23

Config from query
Screenshot from 2021-07-08 19-11-54
Screenshot from 2021-07-08 19-12-27

Prometheus instant query returning threshold:
Screenshot from 2021-07-08 19-12-51

### CODE REVIEW

File: packages/grafana-ui/src/components/Gauge/Gauge.tsx [hunk 1]
```diff
@@ -110,9 +110,11 @@ export class Gauge extends PureComponent<Props> {
     const valueWidthBase = Math.min(width, dimension * 1.3) * 0.9;
     // remove gauge & marker width (on left and right side)
     // and 10px is some padding that flot adds to the outer canvas
-    const valueWidth = valueWidthBase - ((gaugeWidth + (showThresholdMarkers ? thresholdMarkersWidth : 0)) * 2 + 10);
+    const valueWidth =
+      valueWidthBase -
+      ((gaugeWidth + (showThresholdMarkers ? thresholdMarkersWidth : 0) + (showThresholdLabels ? 10 : 0)) * 2 + 10);
     const fontSize = this.props.text?.valueSize ?? calculateFontSize(text, valueWidth, dimension, 1, gaugeWidth * 1.7);
-    const thresholdLabelFontSize = fontSize / 2.5;
+    const thresholdLabelFontSize = Math.max(fontSize / 2.5, 12);
```
Comment (ryantxu) [inline packages/grafana-ui/src/components/Gauge/Gauge.tsx:117]: unrelated?

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  ThresholdsMode,
+} from '@grafana/data';
+
+export interface RowToFieldsTransformOptions {
+  nameField?: string;
+  valueField?: string;
+  mappings?: RowToFieldsTransformMappings[];
+}
+
+export interface RowToFieldsTransformMappings {
```
Comment (ryantxu) [inline public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:23]: Maybe:
```suggestion
export interface FieldToConfigMapping {
```

The row nature is not really true when a reducer is used.

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  fieldName: string;
+  reducerId?: ReducerID;
+  configProperty: string;
```
Comment (ryantxu) [inline public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:26]: Perhaps we could use [keyof FieldConfig] | string?   

For the array based properties (thresholds, etc) we will likely need some better strategy -- maybe threshold[0]?

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  FieldType,
+  getFieldDisplayName,
+  ReducerID,
+  ThresholdsMode,
+} from '@grafana/data';
+
+export interface RowToFieldsTransformOptions {
+  nameField?: string;
+  valueField?: string;
+  mappings?: RowToFieldsTransformMappings[];
```
Comment (ryantxu) [inline public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:20]: is there, or should there be a difference between empty array and undefined?  Perhaps empty means "don't map any config"?  and null will match all known configs?

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+export interface RowToFieldsTransformMappings {
+  fieldName: string;
+  reducerId?: ReducerID;
+  configProperty: string;
+}
+
+export const rowsToFieldsTransformer: DataTransformerInfo<RowToFieldsTransformOptions> = {
+  id: DataTransformerID.rowsToFields,
+  name: 'Rows to fields',
+  description: 'Convert rows to fields with dynamic config',
```
Comment (ryantxu) [inline public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:32]: Will we ever create multiple fields from a single row?  Is there a good use case for that?   If not, perhaps:

```suggestion
  description: 'Convert each row into a field with dynamic config',
```

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+  getFieldDisplayName,
+  ReducerID,
+  ThresholdsMode,
+} from '@grafana/data';
+
+export interface RowToFieldsTransformOptions {
+  nameField?: string;
+  valueField?: string;
+  mappings?: RowToFieldsTransformMappings[];
+}
```
Comment (ryantxu) [inline public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:21]: Maybe:
```suggestion
  labelFields?: string[]
}
```

When not specified, I assume the behavior would be everything not used as name, value, or config gets converted into a label

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+   */
+  operator: (options) => (source) =>
+    source.pipe(
+      map((data) => {
+        // Ignore if we have more than one frame
+        if (data.length !== 1) {
+          return data;
+        }
+
+        return [rowsToFields(options, data[0])];
```
Comment (ryantxu) [inline public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:47]: Why only process one frame?  maybe:

```suggestion
        return data.map( (frame) => rowsToFields(options, frame) );
```

File: packages/grafana-ui/src/components/Gauge/Gauge.tsx (hunk 1 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/Gauge/Gauge.tsx:117]: yea, noticed a Gauge with threshold markers had too big a value, not taking threshold markers into account and noticed threshold markers can become too small to be readable

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts (hunk 2 above)
Comment (torkelo) [reply public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:23]: yea, been thinking about this rename as well :+1:

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts (hunk 3 above)
Comment (torkelo) [reply public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:26]: Well it's not always an exact FieldConfig key, for example threshold1 etc,  yea, maybe threshold[1] but think using only alpha numerical chars could be an advantage to make it easier for query language / syntaxes that might prohibit or require escaping brackets `[ ]`

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts (hunk 4 above)
Comment (torkelo) [reply public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:20]: Why would you want to use this transform if you do not want to map anything?

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts (hunk 6 above)
Comment (torkelo) [reply public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:21]: yea, good a idea

File: public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts (hunk 7 above)
Comment (torkelo) [reply public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts:47]: I guess no reason :)

**Review by dprokop — APPROVED** [review]
This PR is amazing @torkelo, awesome you made it!

I've left a few nits at the current stage, all works as expected. I find the thresholds slightly confusing with the possibility to config only one, but this can be nailed later on.

I'm also just slightly confused with the `Rows to fields` transformer and the fact that it allows config mapping too. What I would expect probably is the `Config from query` to be `Config from frame` and then the `Rows to field` transformer would modify the input frame, and the resulting one could be used as a source of config in the `Config from frame` transformer configured as a separate one.

File: public/app/core/components/TransformersUI/fieldToConfigMapping/fieldToConfigMapping.ts [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+  {
+    key: 'displayName',
+    processor: (value: any) => value.toString(),
+  },
+  {
+    key: 'color',
+    processor: (value) => ({ fixedColor: value, mode: FieldColorModeId.Fixed }),
+  },
+  {
+    key: 'threshold1',
```
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/core/components/TransformersUI/fieldToConfigMapping/fieldToConfigMapping.ts:114]: plan is to add more thresholds (at least 2), Threshold handling is WIP/preview

File: public/app/core/components/TransformersUI/fieldToConfigMapping/fieldToConfigMapping.ts [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+  {
+    key: 'unit',
+    processor: (value) => value.toString(),
+  },
+  {
+    key: 'decimals',
+    processor: toNumericOrUndefined,
+  },
+  {
+    key: 'displayName',
```
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/core/components/TransformersUI/fieldToConfigMapping/fieldToConfigMapping.ts:106]: ah good catch!

**Comment by torkelo** [comment]
> I'm also just slightly confused with the Rows to fields transformer and the fact that it allows config mapping too. What I would expect probably is the Config from query to be Config from frame and then the Rows to field transformer would modify the input frame, and the resulting one could be used as a source of config in the Config from frame transformer configured as a separate one.

The reason it's config from query is the idea is you select a query where config is coming from (select frame by refId). 

The reason it's separate transformations is that Rows to field is actually the more powerful one where it can extract MULTIPLE field configs (each row becomes a field config), and it's mapped to the fields using the name field. This is not possible in `Config from query` it only extracts ONE field config for a full frame (hence it has reducer option) and apply it to other fields using a matcher.  

Rows to fields will have other uses other than extracting config.

File: packages/grafana-data/src/transformations/fieldReducer.ts [hunk 10]
```diff
@@ -232,6 +230,13 @@ export const fieldReducers = new Registry<FieldReducerInfo>(() => [
     description: 'Percentage difference between first and last values',
     standard: true,
   },
+  {
+    id: ReducerID.allValues,
+    name: 'All values',
+    description: 'Returns an array with all values',
+    standard: false,
+    reduce: (field: Field) => ({ allValues: field.values.toArray() }),
```
Comment (ryantxu) [inline packages/grafana-data/src/transformations/fieldReducer.ts:238]: :thinking: we will want to be careful about adding this here -- the general idea for a reducer is that it creates a single value out of many.  By adding this option to the registry, we also add it to everything else that expects a single value.  That may be OK... but need to think on it

Comment (torkelo) [reply packages/grafana-data/src/transformations/fieldReducer.ts:238]: it is a single value :) just an array that is the value. But yes, one idea would be to mark this reducer as "returnsArray" or something so we can optionally include it in StatsPicker only for scenarios when it is supported

File: public/app/core/components/TransformersUI/utils.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import { DataFrame, getFieldDisplayName } from '@grafana/data';
+import { useMemo } from 'react';
+
+export function useAllFieldNamesFromDataFrames(input: DataFrame[]): string[] {
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply public/app/core/components/TransformersUI/utils.ts:4]: Can we reuse: https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/MatchersUI/FieldNameMatcherEditor.tsx#L9

```
const names = useFieldDisplayNames(data);
const selectOptions = useSelectOptions(names, options);
```

?

**Comment by ryantxu** [comment]
Overall looks amazing  :tada:  :heart: 

Maybe "Config from query results" -- since we are using the results, not creating a query?

Can we better separate the config mappings and the apply to?  Perhaps move apply to below the mappings section?

**Review by ryantxu — APPROVED** [review]
I like the new layout much better!

What do you think about changing "Field mappings" to just "Field" -- I worry that we have too many other mapping places already.

image


Do we really need "Apply to self"?  Perhaps it is dangerous, but can we assume that behavior when there is only one frame in the result?
